🚀 4단계 - 리팩터링

nainu opened the pull request:
안녕하세요, 동철님.

앞서 리뷰 감사합니다.
이번에는 리팩토링 작업하고, 코멘트 남겨주신 쿠키 지원도 추가해봤습니다. :)

머지를 하는 바람에 커밋로그가 좀 복잡해졌는데, 실제 커밋은 많지 않습니다.
4단계 붙은 커밋들이랑 제일 마지막 쿠키 지원 추가만 이번에 추가된 커밋입니다. (총 7개)

nainu committed:
docs: README.md 업데이트 (3단계 - 로그인 구현하기)

nainu committed:
feat: 3단계: HTTP Status Code 302

nainu committed:
feat: 3단계: POST 방식으로 회원가입

nainu committed:
feat: 3단계: Cookie에 JSESSIONID 값 저장하기

nainu committed:
feat: 3단계: Session 구현하기

nainu committed:
feat: 3단계: 코드 정리

  - parse 관련 함수를 RequestParser 에서 개별 객체로 옮김.
  - 디버그 로그 추가

nainu committed:
docs: README.md 업데이트 (4단계 - 리팩터링)

nainu committed:
refactor: 4단계: Request* -> HttpRequest*

nainu committed:
refactor: 4단계: HttpResponse 도입

nainu committed:
feat: 4단계: Controller 인터페이스 추가하기

nainu committed:
test: 4단계: HttpResponse 테스트 추가

nainu committed:
refactor: 3단계: 필드/생성자/메서드 순으로 순서 정리

nainu committed:
refactor: 3단계: RequestBody 도입

nainu committed:
refactor: 3단계: getSessionIdFromCookie 를 감추기

nainu committed:
Merge branch 'step3' into step4

nainu committed:
refactor: 3단계: getSession 이 항상 세션을 돌려주도록 수정

nainu committed:
Merge branch 'step3' into step4

nainu committed:
Merge remote-tracking branch 'upstream/nainu' into step4

nainu committed:
refactor: 4단계: HttpResponse 코드 개선

- HttpResponse 로직 리팩토링. ResponseWriter 도입
- MimeTypes 를 enum 으로 변경

nainu committed:
feat: 쿠키 기능 추가

- Controller 코드에서 response.setCookie() 를 통해서 쿠키를 추가할 수 있음
- /?set-cookie-(key)=(value) 호출하여 테스트해볼 수 있음

sah3122 commented on tomcat/src/main/java/camp/nextstep/controller/HelloWorldController.java:
```diff
@@ -0,0 +1,37 @@
+        response.setHeader("HELLO", "World!");
+    }
+
+    private void addCookiesForTesting(HttpResponse response, HttpRequest request) {
```
테스트를 위한 코드는 제거하는것이 좋을것 같아요 😄

nainu replied:
제거 하였습니다 :) 🙇

sah3122 commented on tomcat/src/main/java/camp/nextstep/controller/AbstractController.java:
```diff
@@ -0,0 +1,34 @@
+import javassist.NotFoundException;
+import org.apache.coyote.http11.SessionManager;
+
+public class AbstractController implements Controller {
```
AbstractController 정의 👍

sah3122 commented on tomcat/src/main/java/camp/nextstep/controller/AbstractController.java:
```diff
@@ -0,0 +1,34 @@
+
+public class AbstractController implements Controller {
+
+    private final SessionManager sessionManager;
+
+    public AbstractController() {
+        this.sessionManager = SessionManager.INSTANCE;
+    }
```
Controller 클래스에 SessionManager를 추가해주신 이유가 있을까요 ?

nainu replied:
세션매니저를 싱글턴객체로 사용하기 이전에는 컨트롤러에서 request객체로 세션매니저를 주입해줬었는데, 그때 사용하던 코드입니다. 이제는 필요없어져서 바로 삭제했습니다. :)

sah3122 commented on tomcat/src/main/java/camp/nextstep/controller/LoginController.java:
```diff
@@ -0,0 +1,46 @@
+public class LoginController extends AbstractController {
+
+    @Override
+    protected void doGet(HttpRequest request, HttpResponse response) throws IOException {
```
login controller 구현 깔끔하네요 👍

nainu replied:
인터페이스가 잘 정해져 있어서 고민없이 뽑아낼 수 있었습니다. ㅎㅎ

sah3122 commented on tomcat/src/main/java/camp/nextstep/request/HttpRequest.java:
```diff
@@ -0,0 +1,106 @@
+        return session.getAttribute("user") != null;
+    }
+
+    public void signInAs(User user) throws IOException {
+        final var session = getSession();
+
+        session.setAttribute("user", user);
+        log.debug("로그인: {}", user);
+    }
```
Request 객체가 로그인 기능을 제공하는것이 자연스러울수도, 어색할수도 있을것 같습니다. 
개인적으론 어플리케이션의 기능에 더 가깝다고 생각하는데 동윤님 께서는 어떻게 생각하시나요 ?

nainu replied:
맞습니다. 아까 강의때도 설명해주신 걸 들었는데, HttpRequest 는 org.apache 쪽에, 로그인 로직은 어플리케이션쪽 즉 com.nextstep 패키지 쪽으로 빠져야 될 것 같더라구요. 일단 diff 가 너무 많을 것 같아 패키지 이동은 하지 않았는데, 5단계 미션때 (충돌도 많지 않을 것 같아서) org.apache 쪽으로 일부 클래스들을 좀 옮겨야 할 것 같습니다.

그래서 세션 관련 로직은 HttpRequest 에서 빼서 SessionService 란 곳으로 일단 분리해 냈는데, 더 적당한 위치는 좀더 사용처가 생겨나면 찾아낼 수 있을 것 같습니다.

sah3122 commented on tomcat/src/main/java/camp/nextstep/controller/RequestMapping.java:
```diff
@@ -0,0 +1,21 @@
+    private final RegisterController registerController = new RegisterController();
+    private final StaticPageController staticPageController = new StaticPageController();
+
+    public Controller getController(HttpRequest request) {
```
switch 문을 활용하여 구현해주셨네요 👍 
Map 자료구조의 형태를 가지는것과 분기문으로 제공하는것에 대해 차이점을 고민해보시면 좋을것 같아요 😄

nainu replied:
현재 상태 vs 맵 자료구조 형태가 지금 상황에서 코드 읽기 난이도에는 큰 차이는 없을 것 같습니다.

맵 자료구조 사용했을때 장점이라면 역시 배열대신 맵이 가지는 빠른 탐색속도일 것 같습니다. path 를 맵 키에 매핑시킬 수 있는 문자열로 만들어낼 수 있다면 최적일 것 같습니다.

switch를 이용한 지금 방식의 장점은 매칭에 순서가 있다는 것일 것 같아요.
작성자가 원하는 순서부터 탐색해서 컨트롤러를 돌려줄 수 있는 점이 차별점이 될 것 같아요.
때로는 라우팅끼리 교집합이 생겨서 어느 라우팅이 먼저 발견되냐에 따라 다른 컨트롤러로 요청을 보내버릴 수 있는 상황이 생길 수도 있을텐데 그럴때는 우선순위가 필요할 테니까요.(현재도 staticPageController 가 다른 컨트롤러의 라우팅을 모두 포함하고 있습니다)
나중에 `/user/{username}/detail` 같은 REST 형태의 요청이 들어오면 설정 순서대로 뒤져가면서 매핑될 컨트롤러를 찾아야 할텐데 이 때도 맵 구조보다는 순서가 있는 분기문이 더 어울릴 수 있겠다고,, 생각이 들었습니다. ㅎㅎ

다시보니 맵이 코드는 훨씬 단순하네요 :) 일단 수정은 하지 않아 보았습니다.

```
public class RequestMapping {
    private final StaticPageController DEFAULT_CONTROLLER = new StaticPageController();

    private final Map<String, Controller> routingMap = Map.of(
            "/", new HelloWorldController(),
            "/login", new LoginController(),
            "/register", new RegisterController(),
            "/index.html", new IndexController()
    );

    /**
     * 요구사항: request 객체의 정보를 보고, 처리할 수 있는 컨트롤러를 리턴한다.
     */
    public Controller getController(HttpRequest request) {
        return routingMap.getOrDefault(request.getPath(), DEFAULT_CONTROLLER);
    }
}
```

sah3122 replied:
깊게 고민해주신 흔적이 보이네요 👍 
개인적으로는 복잡한 RequestMapping을 처리하기 위해선 swtich 문을 활용하기 보단 Request만을 처리하기 위한 별도의 클래스를 유지하는것이 유연하게 대처할수 있다고 생각합니다 ! 
현재는 모든 요구사항을 만족하고 있으니 반드시 수정할 필요는 없습니다 💯

sah3122 commented on tomcat/src/main/java/camp/nextstep/response/HttpResponse.java:
```diff
@@ -0,0 +1,117 @@
+import java.util.Map;
+
+public class HttpResponse {
+    private final OutputStream outputStream;
```
HttpResponse 클래스가 Output 관련 책임을 가지는것이 너무 많은 역할을 가졌다고 생각하는데, 응답을 제공하기 위한 클래스로 위임해보는건 어떨까요 ?

nainu replied:
뭔가 찝찝한 느낌이 있었는데 이거였어요. 피드백 감사합니다. 한번 고쳐봤습니다. :)

sah3122 commented on tomcat/src/main/java/camp/nextstep/response/ResponseWriter.java:
```diff
@@ -0,0 +1,69 @@
+
+import static camp.nextstep.response.ResponseWritingStage.*;
+
+public class ResponseWriter {
```
응답을 제공하기 위한 클래스 정의가 인상적이네요 👍

sah3122 reviewed (changes requested):
리팩터링 미션 구현 잘해주셨습니다 👍 
요구사항을 잘 만족하도록 구현해주셨네요 😄 
미션을 끝내기 아쉬워 몇가지 코멘트를 남겨두었는데 확인 부탁드립니다 !

nainu committed:
refactor: 코드 개선(리뷰 반영)

- AbstractController 에서 더이상 필요 없는 세션매니저 관련 코드 제거
- HelloWorldController 코드에서 테스트 코드 제거
- 세션 관련된 로직을 HttpRequest 에서 SessionService 로 이동
- 응답 쓰기 로직을 ResponseWriter 로 위임하도록 수정

nainu committed:
fix: 잘못된 로거 선언 수정

sah3122 commented on tomcat/src/main/java/camp/nextstep/session/SessionService.java:
```diff
@@ -0,0 +1,23 @@
+
+import static java.util.Objects.requireNonNull;
+
+public class SessionService {
```
로그인을 담당하는 비즈니스 로직 분리 👍

sah3122 commented on tomcat/src/main/java/camp/nextstep/response/HttpResponse.java:
```diff
@@ -6,27 +6,28 @@
 import java.util.Map;
 
 public class HttpResponse {
-    private final OutputStream outputStream;
```
피드백 반영 👍

sah3122 reviewed (approved):
피드백 반영 잘해주셨습니다 👍 
마지막 미션까지 화이팅 해주세요 !
